src/grammarconstraints/contains.jl

```diff
@@ -24,3 +24,70 @@ function check_tree(c::Contains, tree::AbstractRuleNode)::Bool
     end
     return any(check_tree(c, child) for child ∈ get_children(tree))
 end
+
+
+"""
+	update_rule_indices!(c::Contains, n_rules::Integer)
+
+Updates a `Contains` constraint to reflect grammar changes. No operation is performed 
+as `Contains` constraints do not require updates for grammar rule changes.
+
+# Arguments
+- `c`: The `Contains` constraint to be updated
+- `n_rules`: The new number of rules in the grammar
+"""
+function HerbCore.update_rule_indices!(c::Contains, n_rules::Integer)
+    # no update required
+end
+
+"""
+	update_rule_indices!(c::Contains, grammar::AbstractGrammar)
+
+Updates a `Contains` constraint to reflect grammar changes. No operation is performed 
+as `Contains` constraints do not require updates for grammar rule changes.
+
+# Arguments
+- `c`: The `Contains` constraint to be updated
+- `grammar`: The grammar that changed
+"""
+function HerbCore.update_rule_indices!(c::Contains, grammar::AbstractGrammar)
+    # no update required
+end
+
+"""
+    update_rule_indices!(c::Contains, n_rules::Integer, mapping::AbstractDict{<:Integer,<:Integer}, constraints::Vector{<:AbstractConstraint})
+
+Updates the `Contains` constraint to reflect grammar changes by replacing it with a new 
+`Contains` constraint using the mapped rule index.
+
+# Arguments
+- `c`: The `Contains` constraint to be updated
+- `n_rules`: The new number of rules in the grammar  
+- `mapping`: Dictionary mapping old rule indices to new rule indices
+- `constraints`: Vector of grammar constraints containing the constraint to update
+"""
+function HerbCore.update_rule_indices!(c::Contains,
+    n_rules::Integer,
+    mapping::AbstractDict{<:Integer,<:Integer},
+    constraints::Vector{<:AbstractConstraint})
+    index = findfirst(x -> x == c, constraints) # assumes no duplicate constraints => TODO: can we assume this?
```

Comment:
```suggestion
    index = only(findall(x -> x == c, constraints))
```

You can make the assumption into an assertion here by using `only(findall(...))`.